### ISSUE TITLE
FT_INVALID_HANDLE

### ISSUE THREAD

**Issue by autoditac** [petersbingham/dae-py-relay-controller#10]
Hi,

using an USB 8 Relay with Windows 10, we get an exception running the test8.py

```
PS C:\Users\username\work\dae-py-relay-controller\Example & Test> .\test8.py
C:\Users\username\work\dae-py-relay-controller\Example & Test\FTD2XX.dll
1093143824
Info1 1093143824,921600
baudrate setting ret val 1
Traceback (most recent call last):
  File "C:\Users\username\work\dae-py-relay-controller\Example & Test\test8.py", line 26, in <module>
    testCommon.TestCommon(dae_RelayBoard.DAE_RELAYBOARD_TYPE_8)
  File "C:\Users\username\work\dae-py-relay-controller\Example & Test\testCommon.py", line 59, in __init__
    dr.initialise(*args)
  File "C:\Python37\lib\site-packages\dae_RelayBoard\dae_RelayBoard.py", line 50, in initialise
    self.relayHandler.initialise(*args)
  File "C:\Python37\lib\site-packages\dae_RelayBoard\dae_RelayBoard_D2XX.py", line 47, in initialise
    self.FTD2XX.initialise(deviceId, self.BAUDRATE, self.MASK, self.BITMODE)
  File "C:\Python37\lib\site-packages\dae_RelayBoard\FTD2XXWindows.py", line 35, in initialise
    self._handleRet(self.FTD2XXDLL.FT_SetBaudRate(self.handle, baudRate))
  File "C:\Python37\lib\site-packages\dae_RelayBoard\FTD2XXWindows.py", line 57, in _handleRet
    raise dae_RelayBoard_Common.Denkovi_Exception(self._getErrString(ret))
dae_RelayBoard.dae_RelayBoard_Common.Denkovi_Exception: FT return error: FT_INVALID_HANDLE 
```
For debugging, we added a print statement to see the handle (2nd line) and the baud rate settings, where it seems to fail (3d line). We added the DLL to our working dir, path is printed in 1st line.

Any idea what might go wrong here and how to debug further?

Cheers,

Rouven

**Comment by petersbingham** [comment]
I'll have a look this weekend. Haven't been near this code for a long time.

I think the dll is being loaded, otherwise you'd have a different error and wouldn't get this far.
Some options:
- check that device is properly enumerated.
- try other usb ports
- that you can initialise using denkovi's software
- that deviceId is correct. (you might be able to get this info from the denkovi SW or from your OS device manager.
- other parameters to the `self.FTD2XX.initialise` function.

**Comment by petersbingham** [comment]
If you check the above and can't identify the issue I'll have a look then. Thanks.

**Comment by noeltaillardat** [comment]
Hi,

I faced the same issue, running the script on Windows with Python3.

I solved it with this little patch:
![image](https://user-images.githubusercontent.com/23383003/101136346-7926e900-360d-11eb-98bf-5f879b5598dc.png)

I didn't try to run it on Linux, but if someone could do it and validate that there is no regression, I hope the patch could be worth a PR.

Regards,

Noël

**Comment by petersbingham** [comment]
Hi there. Thanks a lot for this. I think we can get away without having to test on Linux since this file isn't loaded when using Linux; this only wraps around the windows FDD2XX windows dlls.

However, this may have come about from recent changes to the low level drivers. I'm wondering if this is the case will the c_void_p be backwards compatible with old versions (maybe implicitly cast?). If we can't check this or be sure about it might be worth catching the exception from the old code and executing your fix code then?

**Comment by noeltaillardat** [comment]
The error fixed with those two lines did not raise in `FTD2XXDllWrap`'s `FT_Open`, but when other functions tried to use the handle, so I won't be able to catch something in the opening function :/
*(or I guess I can let previous code, check before returning from Open function by trying any FT function, try to catch something there, and in this case, try to open using my patch ? - it should be a backward compatible option)*

Regarding driver updates, the `handle` returned by Windows `FT_Open` DLL is a void pointer (hidden behind a typedef), and I don't think this has changed. From my point of view, if backward compatible problems might happen, it shall be on Python side (and its use of C types) but I'm definitely not sure of any of those assumptions and won't be able to take time to investigate more (on Python C types handling nor Windows Driver).

**Comment by petersbingham** [comment]
That makes sense and I'm happy with your suggested solution for backwards compatibility. If you submit this with a comment explaining I'll run some tests and will either accept PR or suggest removing the backward compatible code (if tests show it's not required). Hope this sounds ok and thanks for posting this.
________________________________

**Comment by noeltaillardat** [comment]
Hi Peter,

After some tests, it seems that I struggle to implement the "backward-compatible" solution I proposed above: after a FT_Open initializing my handle as something invalid for me, I cannot call the function again (this time with my c_void_p() handle), since I have to FT_Close before ... something that I can't do, since I need a valid handle.

However, during my tests, I discovered that the problem that I had did not happen on one of my remote computers:
I'm not sure if it is because of different driver versions or because this one uses Python 3.8 (the other ones on which current solution failed use latest Python 3.9), but there, the initial code passed. What is interesting, is that my "simple" patch (screenshot above) also worked on this computer ... which represents kind of an example of backward compatibility. It is definitely not exhaustive, but if you have materials to run some tests, I think it's more logical for me to propose that simple two-lines change, as it matches FTD2XX DLL function library (C void pointer used for the handle) and should work whatever Python version is used:

![image](https://user-images.githubusercontent.com/23383003/101385712-743b9100-38bc-11eb-94c3-c468131ad547.png)
*(from https://www.ftdichip.com/Support/Documents/ProgramGuides/D2XX_Programmer's_Guide(FT_000071).pdf)*

and:
![image](https://user-images.githubusercontent.com/23383003/101385543-3a6a8a80-38bc-11eb-95ce-51a38d4757f9.png)
*(from https://www.ftdichip.com/Support/Knowledgebase/index.html?ftd2xx_h.htm)*
